### PR TITLE
fix(router-with-query): context type too narrow

### PR DESCRIPTION
**Pull request by juliusmarminge**
type currently expects exact match on context type:

CleanShot 2024-07-22 at 15 29 35@2x

### CODE REVIEW

File: packages/react-router-with-query/tests/index.test.tsx
Comment (schiller-manuel) [inline packages/react-router-with-query/tests/index.test.tsx:1]: can you please rename to `index.test-d.tsx`

File: packages/react-router-with-query/tests/index.test.tsx [hunk 1]
```diff
@@ -1,5 +1,64 @@
...
+  createRootRouteWithContext,
+  createRouter,
+} from '@tanstack/react-router'
+
+import { expect, expectTypeOf, test } from 'vitest'
+
+import { routerWithQueryClient } from '../src'
 
 test('it has a test', async () => {
   expect(true).toBe(true)
```
Comment (schiller-manuel) [inline packages/react-router-with-query/tests/index.test.tsx:12]: delete this dummy test

File: packages/react-router-with-query/src/index.tsx [hunk 2]
```diff
@@ -4,19 +4,23 @@ import {
...
->(router: T, queryClient: QueryClient): T {
+export function routerWithQueryClient<TRouter extends AnyRouter>(
+  _router: TRouter extends AnyRouterWithContext<infer Ctx>
+    ? Ctx extends { queryClient: QueryClient }
+      ? TRouter
+      : 'Context type mismatch. Expected queryClient to be in context type.'
+    : 'Context type mismatch. Expected queryClient to be in context type.',
+  queryClient: QueryClient,
+): TRouter {
+  const router = _router as unknown as AnyRouter
```
Comment (schiller-manuel) [inline packages/react-router-with-query/src/index.tsx:23]:
```suggestion
export function routerWithQueryClient<TRouter extends AnyRouter>(
  router: TRouter extends AnyRouterWithContext<infer Ctx>
    ? Ctx extends { queryClient: QueryClient }
      ? TRouter
      : never
    : never,
  queryClient: QueryClient,
): TRouter {
```

Comment (schiller-manuel) [reply packages/react-router-with-query/src/index.tsx:23]: when using `never`, you don't have to do any type assertions

File: packages/react-router-with-query/src/index.tsx [hunk 3]
```diff
@@ -123,5 +127,5 @@ export function routerWithQueryClient<
     },
   }
 
-  return router
+  return router as TRouter
```
Comment (schiller-manuel) [inline packages/react-router-with-query/src/index.tsx:130]: not necessary when using `never` above

**Review by schiller-manuel — CHANGES_REQUESTED** [review]

File: packages/react-router-with-query/src/index.tsx (hunk 2 above)
Comment (juliusmarminge) [reply packages/react-router-with-query/src/index.tsx:23]: unfortunately it appears I have to in order to support all the ts versions:
CleanShot 2024-07-23 at 10 04 29

https://github.com/TanStack/router/pull/2007#discussion_r1686586556

Comment (schiller-manuel) [reply packages/react-router-with-query/src/index.tsx]: @chorobin can you please have a look at this?

Comment (tannerlinsley) [reply packages/react-router-with-query/src/index.tsx]: I feel like I’ve run into this before too. Very odd.

Comment (chorobin) [reply packages/react-router-with-query/src/index.tsx]: Fairly sure this is due to invariance. If a type parameter is in a `in` and an `out` position then a router with a different context type is not assignable to `AnyRouterWithContext`

Comment (chorobin) [reply packages/react-router-with-query/src/index.tsx]: Why not
```typescript
export function routerWithQueryClient<TRouter extends AnyRouter>(
  router: TRouter['options']['context'] extends { queryClient: QueryClient }
    ? TRouter
    : AnyRouterWithContext<{ queryClient: QueryClient }>,
  queryClient: QueryClient,
): TRouter
```

Comment (tannerlinsley) [reply packages/react-router-with-query/src/index.tsx]: Well, what do we do? We need this to move forward.

Comment (chorobin) [reply packages/react-router-with-query/src/index.tsx]: I would say my above suggestion works. What do you think @schiller-manuel

Comment (schiller-manuel) [reply packages/react-router-with-query/src/index.tsx]: yes please go for https://github.com/TanStack/router/pull/2007#discussion_r1691905804

Comment (juliusmarminge) [reply packages/react-router-with-query/src/index.tsx]: Will see if it works for all versions

Comment (juliusmarminge) [reply packages/react-router-with-query/src/index.tsx]: CleanShot 2024-07-26 at 23 38 33
 Nope...

**Pull request merged**